### PR TITLE
Implementation of ValidateAndGetBestMatch

### LINKED ISSUE #1699: Rework `CommandService#ExecuteAsync` logic and allow easier extensibility.
In the current state, `ExecuteAsync` does two functions. Command Execution and `Command Validation (Valid command check/ predicate check/ parameter check/ override check, etc.)`.  
My idea is to split those into two separate functions, one responsible for validating the commands and another for execution.
For compatibility and no major breaks, i decided to keep the current public api unchanged with only the addition of a `ValidateAndGetBestMatch` function with the following signature.

```cs
public async Task<(IResult, Optional<CommandMatch>)> ValidateAndGetBestMatch(SearchResult matches,
ICommandContext context,
IServiceProvider provider,
MultiMatchHandling multiMatchHandling = MultiMatchHandling.Exception)
```

This function is to be called with `Search(String)`'s output making `ExecuteAsync`'s logic exclusive for execution,
`Search`'s exclusive for searching and `Validate...` exclusive for validation and command matching.

An *intended* side-effect is the extensibility this would allow if accepted. An external CommandService wrapper could leverage Discord.Net's `Validate...` in order to get the command that best fits **without executing** it, as the only current way of getting the command that **was** called is via an event handler, but there is no clean way for getting a command that *will* be called.

```cs
class Wrapper {
    //Task CheckForWhatever(string)
    var search = _commandService.Search(message.Content.Substring(argPos));
    var (status, command) = await ValidateAndGetBestMatch(search); //Other params ommited for sake of clarity
    if (!status.IsSucess) return;
    command.Value.ExecuteAsync(command.Command, context, status); 
}
```

Status
---
- [x] Implemented on [`siscodeorg/Discord.Net`](https://github.com/siscodeorg/Discord.Net/blob/commands/validate-get-best-match/src/Discord.Net.Commands/CommandService.cs)
- [x] Discussed by community on the discord channel. got mixed reactions.
- [x] PR opened - #1700

### PR DESCRIPTION
**Pull request by alikindsys**
As discussed on #1699 this pr comes with a rewrite of the logic on `CommandService#ExecuteAsync`, splitting it into two functions.    
  
`ExecuteAsync` - will continue to do the `Search-Verify-Match-Execute` loop as it did before.  
  
`ValidateAndGetBestMatch` - Verifies from a set of commands (`SearchResult`) with the specified `ICommandContext`, `IServiceProvider` and `MultiMatchHandling` which command is the optimal one, using the previously established classification and ranking systems. 

Breaking Changes
---
None.

Focus
---
Code cleanup and creation of an new public api.

### CODE REVIEW

**Review by quinchs — APPROVED** [review]

**Comment by quinchs** [comment]
Going to cc @Cenngo on this just for sanity

File: src/Discord.Net.Commands/CommandService.cs
Comment (Cenngo) [inline src/Discord.Net.Commands/CommandService.cs]: I dont think returning an optional CommandMatch to indicate to the user that their SearchResult wasnt successful isnt the most user friendly way of validating a command search. Instead of returning a Tuple, i think you should encapsulate this in an `IResult` that way you can provide more information to the user and it would be easier to build a command execution pipeline around this, since your goal is to allow people to utilize this.

File: src/Discord.Net.Commands/CommandService.cs
Comment (Cenngo) [inline src/Discord.Net.Commands/CommandService.cs]: Tho i understand the way you are using this workflow to branch out the execution method, i dont think it is the most descript way of doing this. I think this should be more user readable to make it easier to maintain. Using an `IResult` to return the method result, with some additional info for branching out the pipeline, might be a better idea.

**Comment by Cenngo** [comment]
Other than that, it all seems good.

File: src/Discord.Net.Commands/CommandService.cs
Comment (alikindsys) [reply src/Discord.Net.Commands/CommandService.cs]: I see.

File: src/Discord.Net.Commands/CommandService.cs
Comment (alikindsys) [reply src/Discord.Net.Commands/CommandService.cs]: Understood.

**Comment by quinchs** [comment]
@roridev Are you able to make the requested changes?

**Comment by alikindsys** [comment]
Im able now, just give me a moment

**Comment by alikindsys** [comment]
Please check if the changes done align with what was requested.

Fun things that were discovered while doing this:

- A whole branch of execution was missing on my old code. The event wouldn't be fired if `ValidateAndGetBestMatch` returned `SearchResult`. This is now fixed.

**Review by quinchs — APPROVED** [review]

**Comment by quinchs** [comment]
@Cenngo waiting on your final approval

File: src/Discord.Net.Commands/CommandService.cs
Comment (Cenngo) [inline src/Discord.Net.Commands/CommandService.cs]: Could you please make this pascal casing?

**Review by Cenngo — CHANGES_REQUESTED** [review]
The changes you made address the issues i mentioned. Great to hear that this also led you to fixing a problem.

File: src/Discord.Net.Commands/CommandService.cs
Comment (alikindsys) [reply src/Discord.Net.Commands/CommandService.cs]: oh sure rider autocorrected it to that.

**Review by Cenngo — APPROVED** [review]

**Review by MrCakeSlayer — APPROVED** [review]

**Pull request merged**
